Heartbeat test failing on CI

Found in https://beats-ci.elastic.co/job/Beats/job/beats/job/PR-25998/8/testReport/junit/github/com_elastic_beats_v7_heartbeat_scheduler_timerqueue/Extended___heartbeat_windows_7_32_windows_7_32_bit___TestQueueRunsInOrder/

Output:

```
=== RUN   TestQueueRunsInOrder
coverage: 81.4% of statements
panic: test timed out after 10m0s

goroutine 105 [running]:
testing.(*M).startAlarm.func1()
	C:/Users/jenkins/workspace/PR-25998-8-5339f7b2-56eb-40bb-b4f1-eb72d047ef3a/.gvm/versions/go1.16.4.windows.386/src/testing/testing.go:1700 +0xcb
created by time.goFunc
	C:/Users/jenkins/workspace/PR-25998-8-5339f7b2-56eb-40bb-b4f1-eb72d047ef3a/.gvm/versions/go1.16.4.windows.386/src/time/sleep.go:180 +0x37

goroutine 1 [chan receive]:
testing.(*T).Run(0x11030fc0, 0x5a4e6a, 0x14, 0x5b3e34, 0xf)
	C:/Users/jenkins/workspace/PR-25998-8-5339f7b2-56eb-40bb-b4f1-eb72d047ef3a/.gvm/versions/go1.16.4.windows.386/src/testing/testing.go:1239 +0x239
testing.runTests.func1(0x11030ee0)
	C:/Users/jenkins/workspace/PR-25998-8-5339f7b2-56eb-40bb-b4f1-eb72d047ef3a/.gvm/versions/go1.16.4.windows.386/src/testing/testing.go:1511 +0x5a
testing.tRunner(0x11030ee0, 0x1105bed0)
	C:/Users/jenkins/workspace/PR-25998-8-5339f7b2-56eb-40bb-b4f1-eb72d047ef3a/.gvm/versions/go1.16.4.windows.386/src/testing/testing.go:1193 +0xc5
testing.runTests(0x110042c0, 0x6d7980, 0x2, 0x2, 0xd315ac6c, 0xc0267b02, 0xb2f62425, 0x8b, 0x6dc8a0, 0x5)
	C:/Users/jenkins/workspace/PR-25998-8-5339f7b2-56eb-40bb-b4f1-eb72d047ef3a/.gvm/versions/go1.16.4.windows.386/src/testing/testing.go:1509 +0x25a
testing.(*M).Run(0x11100190, 0x0)
	C:/Users/jenkins/workspace/PR-25998-8-5339f7b2-56eb-40bb-b4f1-eb72d047ef3a/.gvm/versions/go1.16.4.windows.386/src/testing/testing.go:1417 +0x178
main.main()
	_testmain.go:97 +0x162

goroutine 6 [chan receive]:
github.com/elastic/beats/v7/heartbeat/scheduler/timerqueue.testQueueRunsInOrderOnce(0x11030fc0)
	C:/Users/jenkins/workspace/PR-25998-8-5339f7b2-56eb-40bb-b4f1-eb72d047ef3a/src/github.com/elastic/beats/heartbeat/scheduler/timerqueue/queue_test.go:79 +0x199
github.com/elastic/beats/v7/heartbeat/scheduler/timerqueue.TestQueueRunsInOrder(0x11030fc0)
	C:/Users/jenkins/workspace/PR-25998-8-5339f7b2-56eb-40bb-b4f1-eb72d047ef3a/src/github.com/elastic/beats/heartbeat/scheduler/timerqueue/queue_test.go:33 +0x29
testing.tRunner(0x11030fc0, 0x5b3e34)
	C:/Users/jenkins/workspace/PR-25998-8-5339f7b2-56eb-40bb-b4f1-eb72d047ef3a/.gvm/versions/go1.16.4.windows.386/src/testing/testing.go:1193 +0xc5
created by testing.(*T).Run
	C:/Users/jenkins/workspace/PR-25998-8-5339f7b2-56eb-40bb-b4f1-eb72d047ef3a/.gvm/versions/go1.16.4.windows.386/src/testing/testing.go:1238 +0x21a

goroutine 139 [runnable]:
github.com/elastic/beats/v7/heartbeat/scheduler/timerqueue.(*TimerQueue).Start.func1(0x11184820)
	C:/Users/jenkins/workspace/PR-25998-8-5339f7b2-56eb-40bb-b4f1-eb72d047ef3a/src/github.com/elastic/beats/heartbeat/scheduler/timerqueue/queue.go:74 +0x10a
created by github.com/elastic/beats/v7/heartbeat/scheduler/timerqueue.(*TimerQueue).Start
	C:/Users/jenkins/workspace/PR-25998-8-5339f7b2-56eb-40bb-b4f1-eb72d047ef3a/src/github.com/elastic/beats/heartbeat/scheduler/timerqueue/queue.go:72 +0x49

goroutine 143 [runnable]:
github.com/elastic/beats/v7/heartbeat/scheduler/timerqueue.(*TimerQueue).Start.func1(0x11184960)
	C:/Users/jenkins/workspace/PR-25998-8-5339f7b2-56eb-40bb-b4f1-eb72d047ef3a/src/github.com/elastic/beats/heartbeat/scheduler/timerqueue/queue.go:74 +0x10a
created by github.com/elastic/beats/v7/heartbeat/scheduler/timerqueue.(*TimerQueue).Start
	C:/Users/jenkins/workspace/PR-25998-8-5339f7b2-56eb-40bb-b4f1-eb72d047ef3a/src/github.com/elastic/beats/heartbeat/scheduler/timerqueue/queue.go:72 +0x49

goroutine 145 [runnable]:
github.com/elastic/beats/v7/heartbeat/scheduler/timerqueue.(*TimerQueue).Start.func1(0x11184a00)
	C:/Users/jenkins/workspace/PR-25998-8-5339f7b2-56eb-40bb-b4f1-eb72d047ef3a/src/github.com/elastic/beats/heartbeat/scheduler/timerqueue/queue.go:74 +0x10a
created by github.com/elastic/beats/v7/heartbeat/scheduler/timerqueue.(*TimerQueue).Start
	C:/Users/jenkins/workspace/PR-25998-8-5339f7b2-56eb-40bb-b4f1-eb72d047ef3a/src/github.com/elastic/beats/heartbeat/scheduler/timerqueue/queue.go:72 +0x49

goroutine 179 [select]:
github.com/elastic/beats/v7/heartbeat/scheduler/timerqueue.(*TimerQueue).Start.func1(0x11184aa0)
	C:/Users/jenkins/workspace/PR-25998-8-5339f7b2-56eb-40bb-b4f1-eb72d047ef3a/src/github.com/elastic/beats/heartbeat/scheduler/timerqueue/queue.go:74 +0x10a
created by github.com/elastic/beats/v7/heartbeat/scheduler/timerqueue.(*TimerQueue).Start
	C:/Users/jenkins/workspace/PR-25998-8-5339f7b2-56eb-40bb-b4f1-eb72d047ef3a/src/github.com/elastic/beats/heartbeat/scheduler/timerqueue/queue.go:72 +0x49

goroutine 180 [runnable]:
github.com/elastic/beats/v7/heartbeat/scheduler/timerqueue.(*TimerQueue).Start.func1.1(0x11187fc0, 0xa, 0x10, 0xd3ac1b5c, 0xc0267b02, 0xb3050a99, 0x8b, 0x6dc8a0)
	C:/Users/jenkins/workspace/PR-25998-8-5339f7b2-56eb-40bb-b4f1-eb72d047ef3a/src/github.com/elastic/beats/heartbeat/scheduler/timerqueue/queue.go:82
created by github.com/elastic/beats/v7/heartbeat/scheduler/timerqueue.(*TimerQueue).Start.func1
	C:/Users/jenkins/workspace/PR-25998-8-5339f7b2-56eb-40bb-b4f1-eb72d047ef3a/src/github.com/elastic/beats/heartbeat/scheduler/timerqueue/queue.go:82 +0x19c
```

Feel free to close if not appropriate  / relevant.

I'm seeing this same error with a PR rebased with today's master: https://github.com/elastic/beats/pull/26209

CI job: https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats/detail/PR-26209/13/tests

Test error:
``` 
Stacktrace
=== RUN   TestQueueRunsInOrder
coverage: 79.1% of statements
panic: test timed out after 10m0s
goroutine 162 [running]:
testing.(*M).startAlarm.func1()
	/var/lib/jenkins/workspace/PR-26209-13-ddcab0a4-7c79-45e4-aa83-31ef2705916c/.gvm/versions/go1.16.5.linux.amd64/src/testing/testing.go:1700 +0xe5
created by time.goFunc
	/var/lib/jenkins/workspace/PR-26209-13-ddcab0a4-7c79-45e4-aa83-31ef2705916c/.gvm/versions/go1.16.5.linux.amd64/src/time/sleep.go:180 +0x45
goroutine 1 [chan receive]:
testing.(*T).Run(0xc000106600, 0x607664, 0x14, 0x615bd8, 0x4880c6)
	/var/lib/jenkins/workspace/PR-26209-13-ddcab0a4-7c79-45e4-aa83-31ef2705916c/.gvm/versions/go1.16.5.linux.amd64/src/testing/testing.go:1239 +0x2da
testing.runTests.func1(0xc000106480)
	/var/lib/jenkins/workspace/PR-26209-13-ddcab0a4-7c79-45e4-aa83-31ef2705916c/.gvm/versions/go1.16.5.linux.amd64/src/testing/testing.go:1511 +0x78
testing.tRunner(0xc000106480, 0xc000191da8)
	/var/lib/jenkins/workspace/PR-26209-13-ddcab0a4-7c79-45e4-aa83-31ef2705916c/.gvm/versions/go1.16.5.linux.amd64/src/testing/testing.go:1193 +0xef
testing.runTests(0xc000136138, 0x75bca0, 0x2, 0x2, 0xc02e850d04faf354, 0x8bb2dc17db, 0x765820, 0x6055fc)
	/var/lib/jenkins/workspace/PR-26209-13-ddcab0a4-7c79-45e4-aa83-31ef2705916c/.gvm/versions/go1.16.5.linux.amd64/src/testing/testing.go:1509 +0x2fe
testing.(*M).Run(0xc000188080, 0x0)
	/var/lib/jenkins/workspace/PR-26209-13-ddcab0a4-7c79-45e4-aa83-31ef2705916c/.gvm/versions/go1.16.5.linux.amd64/src/testing/testing.go:1417 +0x1eb
main.main()
	_testmain.go:97 +0x1c5
goroutine 19 [runnable]:
time.NewTimer(0x451093b66f00000, 0x1000)
	/var/lib/jenkins/workspace/PR-26209-13-ddcab0a4-7c79-45e4-aa83-31ef2705916c/.gvm/versions/go1.16.5.linux.amd64/src/time/sleep.go:86 +0xfe
github.com/elastic/beats/v7/heartbeat/scheduler/timerqueue.NewTimerQueue(0x645f78, 0xc000094540, 0x645f78)
	/var/lib/jenkins/workspace/PR-26209-13-ddcab0a4-7c79-45e4-aa83-31ef2705916c/src/github.com/elastic/beats/heartbeat/scheduler/timerqueue/queue.go:50 +0x67
github.com/elastic/beats/v7/heartbeat/scheduler/timerqueue.testQueueRunsInOrderOnce(0xc000106600)
	/var/lib/jenkins/workspace/PR-26209-13-ddcab0a4-7c79-45e4-aa83-31ef2705916c/src/github.com/elastic/beats/heartbeat/scheduler/timerqueue/queue_test.go:44 +0x89
github.com/elastic/beats/v7/heartbeat/scheduler/timerqueue.TestQueueRunsInOrder(0xc000106600)
	/var/lib/jenkins/workspace/PR-26209-13-ddcab0a4-7c79-45e4-aa83-31ef2705916c/src/github.com/elastic/beats/heartbeat/scheduler/timerqueue/queue_test.go:37 +0x35
testing.tRunner(0xc000106600, 0x615bd8)
	/var/lib/jenkins/workspace/PR-26209-13-ddcab0a4-7c79-45e4-aa83-31ef2705916c/.gvm/versions/go1.16.5.linux.amd64/src/testing/testing.go:1193 +0xef
created by testing.(*T).Run
	/var/lib/jenkins/workspace/PR-26209-13-ddcab0a4-7c79-45e4-aa83-31ef2705916c/.gvm/versions/go1.16.5.linux.amd64/src/testing/testing.go:1238 +0x2b3
```

--- a/heartbeat/scheduler/timerqueue/queue.go
+++ b/heartbeat/scheduler/timerqueue/queue.go
@@ -82,7 +82,6 @@ func (tq *TimerQueue) Start() {
 	//if tt == nil {
 	//}
 	go func() {
-		t := time.NewTimer(time.Hour * 999999)
 		for {
 			runtime.Gosched()
 			select {

--- a/heartbeat/scheduler/timerqueue/queue.go
+++ b/heartbeat/scheduler/timerqueue/queue.go
@@ -20,14 +20,7 @@ package timerqueue
 import (
 	"container/heap"
 	"context"
-	"runtime"
 	"time"
-)
-
-const (
-	T_INIT = iota
-	T_STARTED
-	T_CONSUMED
 )
 
 // timerTask represents a task run by the TimerQueue.
@@ -46,7 +39,6 @@ type TimerQueue struct {
 	nextRunAt *time.Time
 	pushCh    chan *timerTask
 	timer     *time.Timer
-	tstate    int
 }
 
 // NewTimerQueue creates a new instance.
@@ -56,7 +48,6 @@ func NewTimerQueue(ctx context.Context) *TimerQueue {
 		ctx:    ctx,
 		pushCh: make(chan *timerTask, 4096),
 		timer:  time.NewTimer(time.Hour * 86400),
-		tstate: T_INIT,
 	}
 	heap.Init(&tq.th)
 
@@ -83,16 +74,11 @@ func (tq *TimerQueue) Start() {
 	//}
 	go func() {
 		for {
-			runtime.Gosched()
 			select {
-			//case <-tt.C:
-			//	panic("WHUT")
-			//fmt.Printf("SLOOOW\n")
 			case <-tq.ctx.Done():
 				// Stop the timerqueue
 				return
 			case now := <-tq.timer.C:
-				tq.tstate = T_CONSUMED
 				tasks := tq.popRunnable(now)
 
 				// Run the tasks in a separate goroutine so we can unblock the thread here for pushes etc.
@@ -105,8 +91,7 @@ func (tq *TimerQueue) Start() {
 				if tq.th.Len() > 0 {
 					nr := tq.th[0].runAt
 					tq.nextRunAt = &nr
-					tq.timer.Reset(nr.Sub(time.Now()))
-					tq.tstate = T_STARTED
+					tq.timer = time.NewTimer(time.Until(nr))
 				} else {
 					tq.timer.Stop()
 					tq.nextRunAt = nil
@@ -122,24 +107,19 @@ func (tq *TimerQueue) pushInternal(tt *timerTask) {
 	heap.Push(&tq.th, tt)
 
 	if tq.nextRunAt == nil || tq.nextRunAt.After(tt.runAt) {
-		//fmt.Printf("T")
-		if tq.tstate == T_STARTED {
-			//fmt.Printf("S")
-			if !tq.timer.Stop() {
-				//fmt.Printf("D")
-				select {
-				case <-tq.timer.C:
-					//fmt.Printf("EXECIT\n")
-					break
-				default:
-				}
-			}
-			//fmt.Printf("X")
-
+		if !tq.timer.Stop() {
+			<-tq.timer.C
 		}
-		tq.timer.Reset(tt.runAt.Sub(time.Now()))
-		tq.tstate = T_STARTED
-
+		// Originally the line below this comment was
+		//   tq.timer.Reset(time.Until(tt.runAt))
+		// however this broke in go1.16rc1, specifically on the commit b4b014465216790e01aa66f9120d03230e4aff46
+		//, specifically on this line:
+		// https://github.com/golang/go/commit/b4b014465216790e01aa66f9120d03230e4aff46#diff-73699b6edfe5dbb3f6824e66bb3566bce9405e9a8c810cac55c8199459f0ac19R652
+		// where some nice new optimizations don't actually work reliably
+		// This can be worked around by instantiating a new timer rather than resetting the timer.
+		// since that internally calls deltimer in runtime/timer.go rather than modtimer,
+		// I suspect that the problem is in modtimer's setting of &pp.timerModifiedEarliest
+		tq.timer = time.NewTimer(time.Until(tt.runAt))
 		tq.nextRunAt = &tt.runAt
 	}
 }

--- a/heartbeat/scheduler/timerqueue/queue.go
+++ b/heartbeat/scheduler/timerqueue/queue.go
@@ -69,9 +69,6 @@ func (tq *TimerQueue) Push(runAt time.Time, fn TimerTaskFn) bool {
 // Start runs a goroutine within the given context that processes items in the queue, spawning a new goroutine
 // for each.
 func (tq *TimerQueue) Start() {
-	//tt := time.NewTimer(time.Second * 5)
-	//if tt == nil {
-	//}
 	go func() {
 		for {
 			select {

--- a/heartbeat/scheduler/timerqueue/queue.go
+++ b/heartbeat/scheduler/timerqueue/queue.go
@@ -104,7 +104,7 @@ func (tq *TimerQueue) pushInternal(tt *timerTask) {
 	heap.Push(&tq.th, tt)
 
 	if tq.nextRunAt == nil || tq.nextRunAt.After(tt.runAt) {
-		if !tq.timer.Stop() {
+		if tq.nextRunAt != nil && !tq.timer.Stop() {
 			<-tq.timer.C
 		}
 		// Originally the line below this comment was

--- a/heartbeat/scheduler/timerqueue/queue.go
+++ b/heartbeat/scheduler/timerqueue/queue.go
@@ -20,7 +20,14 @@ package timerqueue
 import (
 	"container/heap"
 	"context"
+	"runtime"
 	"time"
+)
+
+const (
+	T_INIT = iota
+	T_STARTED
+	T_CONSUMED
 )
 
 // timerTask represents a task run by the TimerQueue.
@@ -39,6 +46,7 @@ type TimerQueue struct {
 	nextRunAt *time.Time
 	pushCh    chan *timerTask
 	timer     *time.Timer
+	tstate    int
 }
 
 // NewTimerQueue creates a new instance.
@@ -48,6 +56,7 @@ func NewTimerQueue(ctx context.Context) *TimerQueue {
 		ctx:    ctx,
 		pushCh: make(chan *timerTask, 4096),
 		timer:  time.NewTimer(time.Hour * 86400),
+		tstate: T_INIT,
 	}
 	heap.Init(&tq.th)
 
@@ -69,13 +78,22 @@ func (tq *TimerQueue) Push(runAt time.Time, fn TimerTaskFn) bool {
 // Start runs a goroutine within the given context that processes items in the queue, spawning a new goroutine
 // for each.
 func (tq *TimerQueue) Start() {
+	//tt := time.NewTimer(time.Second * 5)
+	//if tt == nil {
+	//}
 	go func() {
+		t := time.NewTimer(time.Hour * 999999)
 		for {
+			runtime.Gosched()
 			select {
+			//case <-tt.C:
+			//	panic("WHUT")
+			//fmt.Printf("SLOOOW\n")
 			case <-tq.ctx.Done():
 				// Stop the timerqueue
 				return
 			case now := <-tq.timer.C:
+				tq.tstate = T_CONSUMED
 				tasks := tq.popRunnable(now)
 
 				// Run the tasks in a separate goroutine so we can unblock the thread here for pushes etc.
@@ -89,6 +107,7 @@ func (tq *TimerQueue) Start() {
 					nr := tq.th[0].runAt
 					tq.nextRunAt = &nr
 					tq.timer.Reset(nr.Sub(time.Now()))
+					tq.tstate = T_STARTED
 				} else {
 					tq.timer.Stop()
 					tq.nextRunAt = nil
@@ -104,13 +123,23 @@ func (tq *TimerQueue) pushInternal(tt *timerTask) {
 	heap.Push(&tq.th, tt)
 
 	if tq.nextRunAt == nil || tq.nextRunAt.After(tt.runAt) {
-		// Stop and drain the timer prior to reset per https://golang.org/pkg/time/#Timer.Reset
-		// Only drain if nextRunAt is set, otherwise the timer channel has already been stopped the
-		// channel is empty (and thus would block)
-		if tq.nextRunAt != nil && !tq.timer.Stop() {
-			<-tq.timer.C
+		//fmt.Printf("T")
+		if tq.tstate == T_STARTED {
+			//fmt.Printf("S")
+			if !tq.timer.Stop() {
+				//fmt.Printf("D")
+				select {
+				case <-tq.timer.C:
+					//fmt.Printf("EXECIT\n")
+					break
+				default:
+				}
+			}
+			//fmt.Printf("X")
+
 		}
 		tq.timer.Reset(tt.runAt.Sub(time.Now()))
+		tq.tstate = T_STARTED
 
 		tq.nextRunAt = &tt.runAt
 	}

--- a/heartbeat/scheduler/timerqueue/queue.go
+++ b/heartbeat/scheduler/timerqueue/queue.go
@@ -108,7 +108,9 @@ func (tq *TimerQueue) pushInternal(tt *timerTask) {
 			<-tq.timer.C
 		}
 		// Originally the line below this comment was
-		//   tq.timer.Reset(time.Until(tt.runAt))
+		//
+		//  tq.timer.Reset(time.Until(tt.runAt))
+		//
 		// however this broke in go1.16rc1, specifically on the commit b4b014465216790e01aa66f9120d03230e4aff46
 		//, specifically on this line:
 		// https://github.com/golang/go/commit/b4b014465216790e01aa66f9120d03230e4aff46#diff-73699b6edfe5dbb3f6824e66bb3566bce9405e9a8c810cac55c8199459f0ac19R652

--- a/heartbeat/scheduler/timerqueue/queue_test.go
+++ b/heartbeat/scheduler/timerqueue/queue_test.go
@@ -28,15 +28,17 @@ import (
 )
 
 func TestQueueRunsInOrder(t *testing.T) {
-	t.Skip("flaky test on windows: https://github.com/elastic/beats/issues/26205")
 	// Bugs can show up only occasionally
-	for i := 0; i < 100; i++ {
+	for i := 0; i < 1000000; i++ {
+		if i%1000 == 0 {
+			t.Logf("Runs %d\n", i)
+		}
 		testQueueRunsInOrderOnce(t)
 	}
 }
 
 func testQueueRunsInOrderOnce(t *testing.T) {
-	ctx, ctxCancel := context.WithCancel(context.Background())
+	ctx, ctxCancel := context.WithTimeout(context.Background(), time.Second*5)
 	defer ctxCancel()
 	tq := NewTimerQueue(ctx)
 

--- a/heartbeat/scheduler/timerqueue/queue_test.go
+++ b/heartbeat/scheduler/timerqueue/queue_test.go
@@ -59,9 +59,7 @@ func TestStress(t *testing.T) {
 }
 
 func testQueueRunsInOrderOnce(t *testing.T) {
-	ctx, ctxCancel := context.WithTimeout(context.Background(), time.Second*20)
-	defer ctxCancel()
-	tq := NewTimerQueue(ctx)
+	tq := NewTimerQueue(context.Background())
 
 	// Number of items to test with
 	numItems := 10

--- a/heartbeat/scheduler/timerqueue/queue_test.go
+++ b/heartbeat/scheduler/timerqueue/queue_test.go
@@ -29,11 +29,31 @@ import (
 
 func TestQueueRunsInOrder(t *testing.T) {
 	// Bugs can show up only occasionally
+	var min time.Duration
+	var max time.Duration
+	var slow int
+	var times = []time.Duration{}
 	for i := 0; i < 1000000; i++ {
-		if i%1000 == 0 {
-			t.Logf("Runs %d\n", i)
-		}
+		start := time.Now()
+
 		testQueueRunsInOrderOnce(t)
+
+		duration := time.Now().Sub(start)
+		if duration < min || min == 0 {
+			min = duration
+		}
+		if duration > max {
+			max = duration
+		}
+		if duration > time.Millisecond*5 {
+			slow++
+		}
+		times = append(times, duration)
+
+		if i%1000 == 0 && i > 0 {
+			ninetyFifth := int(float64(len(times)) * 0.95)
+			t.Logf("count: %07d | min: %s\t | max: %s\t| slow: %d\t| @95th: %s\n", i, min, max, slow, times[ninetyFifth])
+		}
 	}
 }
 

--- a/heartbeat/scheduler/timerqueue/queue_test.go
+++ b/heartbeat/scheduler/timerqueue/queue_test.go
@@ -60,8 +60,8 @@ func TestStress(t *testing.T) {
 
 func testQueueRunsInOrderOnce(t *testing.T) {
 	ctx, ctxCancel := context.WithCancel(context.Background())
-	tq := NewTimerQueue(ctx)
 	defer ctxCancel()
+	tq := NewTimerQueue(ctx)
 
 	// Number of items to test with
 	numItems := 10

--- a/heartbeat/scheduler/timerqueue/queue_test.go
+++ b/heartbeat/scheduler/timerqueue/queue_test.go
@@ -19,7 +19,6 @@ package timerqueue
 
 import (
 	"context"
-	"fmt"
 	"math/rand"
 	"os"
 	"runtime/pprof"
@@ -30,71 +29,33 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-func TestQueueRunsInOrder(t *testing.T) {
-	// Bugs can show up only occasionally
-	var min time.Duration
-	var max time.Duration
-	var slow int
-	var failures int
-	for i := 0; i < 1000000; i++ {
-		start := time.Now()
-
-		sd := time.AfterFunc(time.Second*5, func() {
-			t.Logf("Stall detected!")
-			pprof.Lookup("goroutine").WriteTo(os.Stdout, 1)
-		})
-
-		testQueueRunsInOrderOnce(t)
-
-		sd.Stop()
-
-		res := false
-		//res := timerReset()
-		if res {
-			failures++
-		}
-
-		duration := time.Now().Sub(start)
-		if duration < min || min == 0 {
-			min = duration
-		}
-		if duration > max {
-			max = duration
-		}
-		if duration > time.Millisecond*5000 {
-			slow++
-		}
-
-		if i%1000 == 0 && i > 0 {
-			t.Logf("count: %07d | min/max: %s/%s\t| slow: %d\t| fail: %d\n", i, min, max, slow, failures)
-		}
-	}
+func TestRunsInOrder(t *testing.T) {
+	testQueueRunsInOrderOnce(t)
 }
 
-func timerReset() bool {
-	t := time.NewTimer(time.Millisecond)
+// TestStress tries to figure out if we have any deadlocks that show up under concurrency
+func TestStress(t *testing.T) {
+	for i := 0; i < 120000; i++ {
+		failed := make(chan bool)
+		succeeded := make(chan bool)
 
-	ran := []time.Time{}
-	var expired bool
+		watchdog := time.AfterFunc(time.Second*5, func() {
+			failed <- true
+		})
 
-	ctx, _ := context.WithTimeout(context.TODO(), time.Second*5)
+		go func() {
+			testQueueRunsInOrderOnce(t)
+			succeeded <- true
+		}()
 
-	for len(ran) < 10 {
 		select {
-		case now := <-t.C:
-			ran = append(ran, now)
-			if false && !t.Stop() {
-				fmt.Printf("BLOCK\n")
-				<-t.C
-				fmt.Printf("UNB\n")
-			}
-			t.Reset(time.Nanosecond * -2)
-		case <-ctx.Done():
-			expired = true
+		case <-failed:
+			pprof.Lookup("goroutine").WriteTo(os.Stdout, 1)
+			require.FailNow(t, "Scheduler test iteration timed out, deadlock issue?")
+		case <-succeeded:
+			watchdog.Stop()
 		}
 	}
-
-	return expired
 }
 
 func testQueueRunsInOrderOnce(t *testing.T) {

--- a/heartbeat/scheduler/timerqueue/queue_test.go
+++ b/heartbeat/scheduler/timerqueue/queue_test.go
@@ -59,7 +59,9 @@ func TestStress(t *testing.T) {
 }
 
 func testQueueRunsInOrderOnce(t *testing.T) {
-	tq := NewTimerQueue(context.Background())
+	ctx, ctxCancel := context.WithCancel(context.Background())
+	tq := NewTimerQueue(ctx)
+	defer ctxCancel()
 
 	// Number of items to test with
 	numItems := 10

--- a/heartbeat/scheduler/timerqueue/queue_test.go
+++ b/heartbeat/scheduler/timerqueue/queue_test.go
@@ -19,7 +19,10 @@ package timerqueue
 
 import (
 	"context"
+	"fmt"
 	"math/rand"
+	"os"
+	"runtime/pprof"
 	"sort"
 	"testing"
 	"time"
@@ -32,11 +35,24 @@ func TestQueueRunsInOrder(t *testing.T) {
 	var min time.Duration
 	var max time.Duration
 	var slow int
-	var times = []time.Duration{}
+	var failures int
 	for i := 0; i < 1000000; i++ {
 		start := time.Now()
 
+		sd := time.AfterFunc(time.Second*5, func() {
+			t.Logf("Stall detected!")
+			pprof.Lookup("goroutine").WriteTo(os.Stdout, 1)
+		})
+
 		testQueueRunsInOrderOnce(t)
+
+		sd.Stop()
+
+		res := false
+		//res := timerReset()
+		if res {
+			failures++
+		}
 
 		duration := time.Now().Sub(start)
 		if duration < min || min == 0 {
@@ -45,20 +61,44 @@ func TestQueueRunsInOrder(t *testing.T) {
 		if duration > max {
 			max = duration
 		}
-		if duration > time.Millisecond*5 {
+		if duration > time.Millisecond*5000 {
 			slow++
 		}
-		times = append(times, duration)
 
 		if i%1000 == 0 && i > 0 {
-			ninetyFifth := int(float64(len(times)) * 0.95)
-			t.Logf("count: %07d | min: %s\t | max: %s\t| slow: %d\t| @95th: %s\n", i, min, max, slow, times[ninetyFifth])
+			t.Logf("count: %07d | min/max: %s/%s\t| slow: %d\t| fail: %d\n", i, min, max, slow, failures)
 		}
 	}
 }
 
+func timerReset() bool {
+	t := time.NewTimer(time.Millisecond)
+
+	ran := []time.Time{}
+	var expired bool
+
+	ctx, _ := context.WithTimeout(context.TODO(), time.Second*5)
+
+	for len(ran) < 10 {
+		select {
+		case now := <-t.C:
+			ran = append(ran, now)
+			if false && !t.Stop() {
+				fmt.Printf("BLOCK\n")
+				<-t.C
+				fmt.Printf("UNB\n")
+			}
+			t.Reset(time.Nanosecond * -2)
+		case <-ctx.Done():
+			expired = true
+		}
+	}
+
+	return expired
+}
+
 func testQueueRunsInOrderOnce(t *testing.T) {
-	ctx, ctxCancel := context.WithTimeout(context.Background(), time.Second*5)
+	ctx, ctxCancel := context.WithTimeout(context.Background(), time.Second*20)
 	defer ctxCancel()
 	tq := NewTimerQueue(ctx)
 

